README
NAME
    Path::Hilbert::XS - XS implementation of a Hilbert Path algorithm

SYNOPSIS
        use Path::Hilbert::XS;

        my ($x, $y) = d2xy(16, 127);
        my $d = xy2d(16, $x, $y);
        die unless $d == 127;

DESCRIPTION
    This implements Path::Hilbert in XS for speed and awesomesauceness.

    The OO interface is not available (yet?).

WHY
    While Path::Hilbert is a solid module, we just wanted speed when dealing
    with abundant amount of data.

    Here are some statistics, generated using the bundled tools/benchmark.pl
    script:

        -- d2xy --
        PP: Rounded run time per iteration: 1.4496e-05 +/- 4.8e-08 (0.3%)
        XS: Rounded run time per iteration: 5.1150e-07 +/- 6.3e-10 (0.1%)

            Time       PP      XS
        PP  1.450e-05  --      -3.53%
        XS  5.115e-07  96.47%  --

        -- xy2d--
        PP: Rounded run time per iteration: 1.3215e-05 +/- 2.7e-08 (0.2%)
        XS: Rounded run time per iteration: 4.7877e-07 +/- 8.3e-10 (0.2%)

            Time       PP      XS
        PP  1.322e-05  --      -3.62%
        XS  4.788e-07  96.38%  --

    (Statistics collected using Dumbbench.)

                Rate    PP    XS
        PP   71028/s    --  -98%
        XS 3703704/s 5114%    --

    (Statistics collected using Benchmark.)

CREDITS
    *   Rafaël Garcia-Suarez - for asking for it.

    *   p5pclub

AUTHORS
    *   Gonzalo Diethelm "gonzus AT cpan DOT org"

    *   Sawyer X "xsawyerx AT cpan DOT org"



Changes
0.002   2015-08-12
        * Improve portability. (Rafael Garcia Suarez)
        * Added this Changelog.
        * Some statistics in POD.

0.001   2015-08-10
        * First version.


Makefile.PL
use 5.008005;
use ExtUtils::MakeMaker;
WriteMakefile(
    NAME           => 'Path::Hilbert::XS',
    VERSION_FROM   => 'lib/Path/Hilbert/XS.pm',
    PREREQ_PM      => { 'Test::More' => 0 },
    ABSTRACT_FROM  => 'lib/Path/Hilbert/XS.pm',
    AUTHOR         => [
        'Sawyer X <xsawyerx@cpan.org>',
        'Gonzalo Diethelm <gonzus@cpan.org>',
    ],
    LICENSE        => 'mit',
    DEFINE         => '',
    INC            => '-I.',
    OBJECT         => '$(O_FILES)',
    META_MERGE     => {
        'meta-spec' => { version => 2 },
        resources   => {
            repository => {
                type => 'git',
                url  => 'git://github.com/p5pclub/path-hilbert-xs.git',
                web  => 'https://github.com/p5pclub/path-hilbert-xs',
            },
        },
    },
);

# Add -Wall to CCFLAGS without overwriting the defaults
package MY;
sub cflags {
    my $self   = shift;
    my $cflags = $self->SUPER::cflags(@_);
    my @cflags = split /\n/, $cflags;

    $self->{CCFLAGS} .= ' -Wall ';
    foreach (@cflags) {
        $_ = "CCFLAGS = $self->{CCFLAGS}" if /^CCFLAGS/;
    }
    return $self->{CFLAGS} = join("\n", @cflags) . "\n";
}


tools/benchmark.pl
#!/usr/bin/perl
use strict;
use warnings;
use Dumbbench;
use Path::Hilbert ();
use Path::Hilbert::XS ();
use Benchmark::Dumb ':all';
use Text::Table;
use DDP;

sub show_statistics_table {
    my $bench = shift;
    my @instances = @{ $bench->{'instances'} };
    my @names     = map $_->name, @instances;
    my $table     = Text::Table->new( '', ' Time', map " $_", @names );

    foreach my $idx ( 0 .. $#instances ) {
        my $my_cb_name = $instances[$idx]{'name'};
        my $my_cb_num  = $instances[$idx]{'result'}{'num'};
        my @items      = ( $my_cb_name, sprintf ' %.3e', $my_cb_num );

        foreach my $cmp_idx ( 0 .. $#instances ) {
            my $cmp_cb_name = $instances[$cmp_idx]{'name'};
            my $cmp_cb_num  = $instances[$cmp_idx]{'result'}{'num'};

            push @items, $my_cb_name eq $cmp_cb_name
                       ? ' --'
                       : sprintf " %.2f%%\n",
                        (
                           $my_cb_num <= $cmp_cb_num
                           ? 100 - $my_cb_num / $cmp_cb_num * 100
                           : $cmp_cb_num / $my_cb_num * -100
                        );
        }

        $table->load([@items]);
    }

    print $table;
}

sub run_instances {
    my @instances = @_;
    my $bench     = Dumbbench->new(
        target_rel_precision => 0.005,
        initial_runs         => 20,
    );

    $bench->add_instances(@_);
    $bench->run;
    $bench->report;
    show_statistics_table($bench);
}

print "-- d2xy --\n";
run_instances(
    Dumbbench::Instance::PerlSub->new(
        name => 'PP',
        code => sub { Path::Hilbert::d2xy( 16, 127 ) for 1e8 },
    ),

    Dumbbench::Instance::PerlSub->new(
        name => 'XS',
        code => sub { Path::Hilbert::XS::d2xy( 16, 127 ) for 1e8 },
    ),
);

print "\n-- xy2d--\n";

run_instances(
    Dumbbench::Instance::PerlSub->new(
        name => 'PP',
        code => sub { Path::Hilbert::xy2d( 16, 7, 8 ) for 1e8 },
    ),

    Dumbbench::Instance::PerlSub->new(
        name => 'XS',
        code => sub { Path::Hilbert::XS::xy2d( 16, 7, 8 ) for 1e8 },
    ),
);


lib/Path/Hilbert/XS.pm
package Path::Hilbert::XS;
use strict;
use warnings;
use XSLoader;
use parent 'Exporter';

our $VERSION = '0.002';
our @EXPORT  = qw<d2xy xy2d>;

XSLoader::load( 'Path::Hilbert::XS', $VERSION );

1;

__END__

=encoding utf8

=head1 NAME

Path::Hilbert::XS - XS implementation of a Hilbert Path algorithm

=head1 VERSION

0.002

=head1 SYNOPSIS

    use Path::Hilbert::XS;

    my ($x, $y) = d2xy(16, 127);
    my $d = xy2d(16, $x, $y);
    die unless $d == 127;

=head1 DESCRIPTION

This implements L<Path::Hilbert> in XS for speed and awesomesauceness.

The OO interface is not available (yet?).

=head1 WHY

While L<Path::Hilbert> is a solid module, we just wanted speed when dealing
with abundant amount of data.

Here are some statistics, generated using the bundled F<tools/benchmark.pl>
script:

    -- d2xy --
    PP: Rounded run time per iteration: 1.4496e-05 +/- 4.8e-08 (0.3%)
    XS: Rounded run time per iteration: 5.1150e-07 +/- 6.3e-10 (0.1%)

        Time       PP      XS
    PP  1.450e-05  --      -3.53%
    XS  5.115e-07  96.47%  --

    -- xy2d--
    PP: Rounded run time per iteration: 1.3215e-05 +/- 2.7e-08 (0.2%)
    XS: Rounded run time per iteration: 4.7877e-07 +/- 8.3e-10 (0.2%)

        Time       PP      XS
    PP  1.322e-05  --      -3.62%
    XS  4.788e-07  96.38%  --

(Statistics collected using L<Dumbbench>.)

            Rate    PP    XS
    PP   71028/s    --  -98%
    XS 3703704/s 5114%    --

(Statistics collected using L<Benchmark>.)

=head1 CREDITS

=over 4

=item * Rafaël Garcia-Suarez - for asking for it.

=item * p5pclub

=back

=head1 AUTHORS

=over 4

=item * Gonzalo Diethelm C<< gonzus AT cpan DOT org >>

=item * Sawyer X C<< xsawyerx AT cpan DOT org >>

=back

